### PR TITLE
Add check if all the packages are listed in the `aea-config.yaml`

### LINKED ISSUE #254: Might not be an issue - needs checking
**Is your feature request related to a problem? Please describe.**
I'm not sure if it is present or not.

**Describe the solution you'd like**
When loading aea-config.yaml, we should ensure that it references ALL the package dependencies, even nested ones. So that `aea-config.yaml` is always flat.

The non-agent packages should always be non-flat: https://github.com/valory-xyz/open-aea/issues/253

### PR DESCRIPTION
**Pull request by angrybayblade**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

## Fixes

#254 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/data/dummy_aea/aea-config.yaml [hunk 1]
```diff
@@ -18,6 +18,7 @@ protocols:
 - open_aea/signing:1.0.0
 skills:
 - dummy_author/dummy:0.1.0
+- dummy_author/test_skill:0.1.0
```
Comment (angrybayblade) [inline tests/data/dummy_aea/aea-config.yaml:21]: I scaffolded this skill for the test.

**Review by Karrenbelt — APPROVED** [review]
LGTM

**Comment by 8ball030** [comment]
LGTM

File: aea/aea_builder.py [hunk 2]
```diff
@@ -1644,6 +1645,33 @@ def _check_pypi_dependencies(self, configuration: ComponentConfiguration) -> Non
...
+    ) -> None:
+        """Check project config for missing dependencies."""
+
+        dep_tree: Set[ComponentId] = set()
+        for level in DependencyTree.generate(project_path):
+            dep_tree.update(
+                {
+                    ComponentId(package.package_type.value, package.public_id)
+                    for package in level
+                    if package.package_type != PackageType.AGENT
```
Comment (DavidMinarsch) [inline aea/aea_builder.py:1660]: It's impossible to be a package type agent. If it was, we should raise a ValueError imo as it indicates a problem in our code.

File: aea/aea_builder.py [hunk 3]
```diff
@@ -1644,6 +1645,33 @@ def _check_pypi_dependencies(self, configuration: ComponentConfiguration) -> Non
...
+                }
+            )
+
+        available_components = {
+            component.without_hash()
+            for component in agent_configuration.all_components_id
+        }
+
+        missing_dependencies_from_config = dep_tree - available_components
```
[the remark this answers is not in the document]
Comment (angrybayblade) [reply aea/aea_builder.py:1669]: But we already have a check in place for that situation, eg. if a packages is listed in the config but missing as a package the build will fail.

File: aea/aea_builder.py (hunk 2 above)
Comment (angrybayblade) [reply aea/aea_builder.py:1660]: So the dependency tree generator scans for all available packages in the project dir which means it'll also count in the agent package so we have to filter it out.

**Review by DavidMinarsch — APPROVED** [review]

File: aea/helpers/dependency_tree.py [hunk 4]
```diff
@@ -198,6 +198,8 @@ def generate(cls, packages_dir: Path) -> List[List[PackageId]]:
                     "name", item_config.get("agent_name")
                 )
                 public_id = PublicId.from_json(item_config)
+                if public_id.name == "scaffold":
+                    continue
```
Comment (DavidMinarsch) [inline aea/helpers/dependency_tree.py:202]: This is not the right place to ignore. The generate method's responsibility is neatly defined and should not have this custom exception.

File: aea/helpers/dependency_tree.py [hunk 5]
```diff
@@ -192,7 +193,16 @@ def generate(cls, packages_dir: Path) -> List[List[PackageId]]:
         """Returns PublicId to hash mapping."""
         package_to_dependency_mappings = {}
         for component_type, component_file in COMPONENTS:
-            for file_path in packages_dir.glob(f"**/{component_file}"):
+            packages_list = []
+            if (packages_dir / VENDOR).exists():
+                packages_list = [
+                    *Path(".", VENDOR).glob(f"**/{component_file}"),
+                    *Path(".", f"{component_type}s").glob(f"**/{component_file}"),
+                ]
```
Comment (angrybayblade) [inline aea/helpers/dependency_tree.py:201]: When scanning for packages in an aea project, we only scan in `skills, connections, protocols, contracts` and `vendor` rather then scanning for whole directory.

File: aea/helpers/dependency_tree.py [hunk 6]
```diff
@@ -187,20 +189,63 @@ def flatten_tree(
...
+    def find_packages_in_a_project(
+        project_dir: Path,
+    ) -> List[Tuple[str, Path]]:
+        """Find packages in an AEA project."""
+
+        packages = []
+        for package_type, config_file in COMPONENTS:
+            _packages = [
+                *Path(project_dir, VENDOR).glob(f"*/{package_type}s/*/{config_file}"),
+                *Path(project_dir, f"{package_type}s").glob(f"*/{config_file}"),
```
Comment (angrybayblade) [inline aea/helpers/dependency_tree.py:202]: So I've added more strict glob patterns, think this should address your concerns @DavidMinarsch

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**